### PR TITLE
Remove `--with` and `--without` flags

### PR DESCRIPTION
**Pull request by smlance**

### CODE REVIEW

File: lib/bundler/cli/config.rb [hunk 1]
```diff
@@ -92,5 +87,147 @@ def run
...
+      elsif name == "path" and Bundler.settings["path.system"]
+        Bundler.ui.warn "`path.system` is already configured, so it will be unset."
+        delete_config("path.system")
+        :conflict
+      else
+        :no_conflict
+      end
+    end
+
+    def resolve_with_without_conflicts(name, new_value, scope = "global")
```
Comment (segiddins) [inline lib/bundler/cli/config.rb]: `resolve_group_conflicts` ?

File: lib/bundler/cli.rb [hunk 2]
```diff
@@ -276,6 +276,11 @@ def package
       bundle exec you can require and call the bundled gems as if they were installed
       into the system wide Rubygems repository.
     D
+    method_option "local", :type => :boolean, :default => false, :banner =>
```
Comment (segiddins) [inline lib/bundler/cli.rb]: look like they're in the wrong place

File: lib/bundler/cli/config.rb [hunk 3]
```diff
@@ -13,11 +13,16 @@ def run
...
+        # The format is `bundle config --local name ...`
         name = args.shift
         scope = $'
       else
+        # The format is `bundle config name ...`
         name = peek
-        scope = "global"
+        # If setting `with` or `without`, allow the scope to be specified anywhere in the command.
+        # Otherwise, set the scope to "global".
+        simple_args = (args.find {|arg| arg =~ /^\-\-/ })
```
Comment (segiddins) [inline lib/bundler/cli/config.rb]: `start_with?`

File: lib/bundler/cli/config.rb [hunk 4]
```diff
@@ -94,5 +98,99 @@ def run
...
+        Bundler.ui.warn "`path.system` is already configured, so it will be unset."
+        delete_config("path.system")
+        :conflict
+      else
+        :no_conflict
+      end
+    end
+
+    def resolve_with_without_conflicts(name, new_value, scope = "global")
+      new_value = new_value.gsub("--global", "").gsub("--local", "").strip # FIXME: This is unnecessary
```
Comment (segiddins) [inline lib/bundler/cli/config.rb]: fix it?

File: lib/bundler/cli/config.rb [hunk 5]
```diff
@@ -94,5 +98,99 @@ def run
...
+        end
+
+        :conflict
+      elsif (name == "without") && with_conflict?(groups, scope)
+        with_scope = groups_conflict?(:with, groups, :local, scope) ? "locally" : "globally"
+        conflicts = conflicting_groups(:with, groups, with_scope == "locally" ? :local : :global, scope)
+
+        Bundler.ui.info "`with` and `without` settings cannot share groups. "\
+         "You have already set `with #{conflicts.join(" ")}` #{with_scope}, so it will be unset."
+        # Bundler.settings.with = Bundler.settings.with - [group]
```
Comment (segiddins) [inline lib/bundler/cli/config.rb]: please don't leave commented-out code

File: lib/bundler/cli/config.rb [hunk 6]
```diff
@@ -94,5 +98,99 @@ def run
...
+    # end
+
+    # group_conflict?: Detects conflicts in optional groups in consideration of scope.
+    # - `name` is the option name (`with` or `without`).
+    # - `groups` is an array of the name(s) of the included or excluded group(s).
+    # - `scope_prev` is the scope of the option previously set.
+    # - `scope_new` is the scope of the option the user is currently trying to set.
+    # NOTE: scope_prev and scope_new must be local or global.
+    def groups_conflict?(name, groups, scope_prev, scope_new)
+      int = conflicting_groups(name, groups, scope_prev, scope_new)
```
Comment (segiddins) [inline lib/bundler/cli/config.rb]: please rename this variable...

File: lib/bundler/cli/config.rb [hunk 7]
```diff
@@ -94,5 +98,99 @@ def run
...
+    def delete_config(name, scope = nil)
+      Bundler.settings.set_local(name, nil) unless scope == "global"
+      Bundler.settings.set_global(name, nil) unless scope == "local"
+    end
+
+    # def group_conflict?(name, group, scope_prev, scope_new)
+    #   Bundler.settings.send(name.to_sym, scope_prev).include?(group) && scope_new.to_sym == scope_prev
+    # end
+
+    # group_conflict?: Detects conflicts in optional groups in consideration of scope.
```
Comment (segiddins) [inline lib/bundler/cli/config.rb]: please use proper YARD doc comments

File: lib/bundler/settings.rb [hunk 8]
```diff
@@ -125,17 +125,31 @@ def with=(array)
...
 
-    def without
-      get_array(:without)
+    def without(scope = nil)
+      key = key_for(:without)
+      if scope.nil?
+        get_array(:without)
+      elsif scope == :global
+        @global_config[key] ? @global_config[key].split(" ").map(&:to_sym) : []
+      elsif scope == :local
```
Comment (segiddins) [inline lib/bundler/settings.rb]: what if the scope is something else?

**Comment by smlance** [comment]
I'll add documentation tomorrow.

**Comment by indirect** [comment]
looking good. @homu r+

**Pull request merged**